ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+                // perform in a single step. Since this is such a critical step, and since we're
+                // prepared to bail out early, there's no point in yielding: either we win the race
+                // by trying repeatedly or we give up entirely.
+                let hl_ptr = Box::into_raw(Box::new(HotLocation::Tracing(None)));
+                let new_ls = State::new().with_hotlocation(hl_ptr).with_lock();
                 loop {
-                    let old_lp = lp;
-                    lp = loc.compare_and_swap(lp, State::phase_tracing_lock(), Ordering::Acquire);
-                    if lp == old_lp {
-                        break;
-                    }
-                    if self.inner.tracer.is_none() {
-                        // This thread isn't tracing anything, so it's not worth us spinning and
-                        // trying to lock this location.
-                        return None;
-                    }
-                    // We are tracing something, and since it might be the current Location, we
-                    // need to transition to TRACING_LOCK to avoid tracing more than one iteration
-                    // of the loop. However, we might be able to determine that this thread
-                    // couldn't have been tracing this location, at which point we can bail out.
-                    // FIXME: this is a terrible spinlock.
-                    loop {
-                        match lp.phase() {
-                            PHASE_TRACING => break,
-                            PHASE_TRACING_LOCK => (),
-                            PHASE_COMPILED | PHASE_COMPILING | PHASE_LOCKED | PHASE_COUNTING
-                            | PHASE_DONT_TRACE => {
-                                // If we observe any of these states, then another thread took
-                                // responsibility for this Location, proving that this thread can't
-                                // be responsible for it.
+                    debug_assert!(!ls.is_locked());
+                    match loc.compare_exchange_weak(
+                        ls,
+                        new_ls,
+                        Ordering::Acquire,
+                        Ordering::Relaxed,
+                    ) {
+                        Ok(_) => {
+                            // We've initialised this Location and obtained the lock, so we can now
+                            // start tracing for real.
+                            let tid = Arc::clone(&self.inner.tid);
+                            let tt = start_tracing(self.inner.tracing_kind);
+                            *unsafe {new_ls.hot_location()} = HotLocation::Tracing(Some((tid, tt)));
+                            Rc::get_mut(&mut self.inner).unwrap().tracing =
+                                Some(hl_ptr as *const ());
+                            loc.unlock();
+                            return None;
+                        }
+                        Err(x) => {
```

Comment:
You do love to use `x` as a variable in this PR. :stuck_out_tongue: I can guess it's a `State`, but it's not super clear. Maybe name the variable `state`?

Reply:
It could be argued this should be named `ls` but that would conflict with the `ls` in an outer scope. Naming is a tar pit!